### ISSUE TITLE
[expo-image] Memory leak when rendering gifs on android

### ISSUE THREAD

**Issue by terrysahaidak** [expo/expo#24259]
### Minimal reproducible example

https://github.com/terrysahaidak/expo-image-gifs-memory-leak

### Summary

Implementing a feature with reactions in the app, I noticed random crashes. After investigating further, i found out that crashes are happening because of memory leak in glide while rendering gifs.

I implemented a repo where you can show-hide gifs until the app crashes (it takes only 5 times for this)
It does not happen on iOS.

The memory graph looks like this:
<img width="1788" alt="image" src="https://github.com/expo/expo/assets/7809008/a8cb0150-4a31-483a-9593-1d8ab2a53ae8">

Here is the crash error:

<details>
<summary>Details</summary>

```

FATAL EXCEPTION: FrameDecoderExecutor-0
                                                                                                    Process: com.expoimagegifsleak, PID: 24183
                                                                                                    java.lang.OutOfMemoryError: Failed to allocate a 921616 byte allocation with 501896 free bytes and 490KB until OOM, target footprint 268435456, growth limit 268435456
                                                                                                    	at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:54)
                                                                                                    	at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:49)
                                                                                                    	at java.nio.ByteBuffer.allocate(ByteBuffer.java:284)
                                                                                                    	at com.github.penfeizhou.animation.decode.FrameSeqDecoder.initCanvasBounds(FrameSeqDecoder.java:250)
                                                                                                    	at com.github.penfeizhou.animation.decode.FrameSeqDecoder.access$700(FrameSeqDecoder.java:36)
                                                                                                    	at com.github.penfeizhou.animation.decode.FrameSeqDecoder$5.run(FrameSeqDecoder.java:233)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:942)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.os.HandlerThread.run(HandlerThread.java:67)
```

</details>

### Environment

```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 16.15.1 - ~/.nvm/versions/node/v16.15.1/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v16.15.1/bin/yarn
      npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
      Watchman: 2023.08.14.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /Users/terrysahaidak/.rvm/gems/ruby-2.7.5/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
      Android SDK:
        API Levels: 28, 29, 30, 31, 32, 33
        Build Tools: 28.0.3, 30.0.2, 30.0.3, 31.0.0, 32.0.0, 33.0.0
        System Images: android-30 | Google APIs ARM 64 v8a, android-30 | Google Play ARM 64 v8a, android-32 | Google Play ARM 64 v8a, android-33 | Google Play ARM 64 v8a, android-Tiramisu | Google APIs ARM 64 v8a
        Android NDK: 21.3.6528147
    IDEs:
      Android Studio: 2022.2 AI-222.4459.24.2221.9971841
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmPackages:
      expo: ^49.0.0 => 49.0.9 
      react: 18.2.0 => 18.2.0 
      react-native: 0.72.4 => 0.72.4 
    npmGlobalPackages:
      eas-cli: 3.13.2
    Expo Workflow: bare
```

**Comment by RRaideRR** [comment]
super interesting thing. [I also fight with memory leaks on Android](https://github.com/expo/expo/issues/22438). I always thought it is the case because I'm having horizontal FlashLists stacked in a vertical one and this is just too much for Android.

Does it only happens with GIFs or do you also experience this with other image formats (or other animated images like webp)?

**Comment by terrysahaidak** [comment]
I didn't try webp, but it does happen only for gifs and only on Android.

I think it doesn't release the gifs from native. But also would be cool if it could use the caching prop and actually keep in memory these gifs since I have a single set I use a lot. But right now it just loads them from scratch all the time.

**Comment by terrysahaidak** [comment]
WebP leaks the same as gif, just slower due to smaller size.

Also, FastImage doesn't have this problem.

**Comment by terrysahaidak** [comment]
After some more investigation, i realized that the only huge change between fast-image and expo-image implementations on Android is the library that is used for animating gifs

Commenting out this line fixes the memory leak and crash but the gifs won't animate anymore:
https://github.com/expo/expo/blob/main/packages/expo-image/android/build.gradle#L103

**Comment by RRaideRR** [comment]
Doesn't this contradicts the assumption that the memory leak also occurs with WebP images?

**Comment by terrysahaidak** [comment]
> Doesn't this contradicts the assumption that the memory leak also occurs with WebP images?

The memory leak was happening for animated webp which is animated by that library too.

**Comment by jingpeng** [comment]
Hi, 
we found the implementation of expo-image only calls the start method of the drawable but not call stop method, so we add a new pr to implements when glide recycle the drawable as follows:

![image](https://github.com/expo/expo/assets/5223226/0f0fe351-c4ad-4cf9-bedc-e6a3997e7a1c)

https://github.com/penfeizhou/APNG4Android/pull/204

Please update to 

`implementation "com.github.penfeizhou.android.animation:glide-plugin:2.28.0"`

fix the memory leak

**Comment by devoren** [comment]
@RRaideRR hey! in my case it also lags and fps around 5-10fps inside flashlist, i use simple png, jpg formats from cdn

**Comment by terrysahaidak** [comment]
> @RRaideRR hey! in my case it also lags and fps around 5-10fps inside flashlist, i use simple png, jpg formats from cdn

could be related since when the new source is used the old gif is never released

**Comment by devoren** [comment]
This also happens outside of the flashlist when I just use it.

**Comment by tsapeta** [comment]
#24358 solved the issue. I've just published this change in `expo-image@1.5.1`, please let me know if it works!
